### ISSUE TITLE
SignClient.Session.Keys NullReferenceException on Signature Request

### ISSUE THREAD

**Issue by underclockeddev** [WalletConnect/WalletConnectUnity#173]
**Describe the bug**
When trying to make this call to get a signature from user's wallet: 
await WalletConnect.Instance.RequestAsync<WCEthSignTypedDataV4, string>(data);

This is sometimes the state of the instance:
WalletConnect.Instance.ActiveSession.Key = "123...890"
WalletConnect.Instance.ActiveSession.Topic = "123...890"
WalletConnect.Instance.SignClient.Session.Keys = null

which causes WalletConnectSharp.Sign.Engine.IsValidSessionTopic("123...890") to throw WalletConnectException.FromType(ErrorType.NO_MATCHING_KEY, "session topic doesn't exist 123...890");

**To Reproduce**
1. Connect to WalletConnectModal 
2. Click button which should request signature
3. NullReferenceException is thrown when trying to read WalletConnect.Instance.SignClient.Session.Keys

```
//this is my SignPermit logic (which definitely does work, when the SignClient.Session.Keys is not null)
protected async Task<string> SignPermit(Permit permit)
{
    TypedData<Domain> typedData = BuildPayload(permit);
    Web3 web3 = new Web3();
    var service = new NethereumWalletConnectService(WalletConnect.Instance.SignClient);
    web3.Client.OverridingRequestInterceptor = new NethereumWalletConnectInterceptor(service);
    string response = string.Empty;
    try
    {
        string payload = typedData.ToJson();
        WCEthSignTypedDataV4 data = new WCEthSignTypedDataV4(Address, payload);
        response = await web3.Eth.AccountSigning.SignTypedDataV4.SendRequestAsync(payload); //exception is thrown within this call
        ......
    }
}
```

**Expected behavior**
1. Keys is never null and the signature is requested each time the button is clicked

**Environment**
 - OS: EndeavourOS Linux
 - Unity version: 2021.3.18f1
 - Packages: 
     - WalletConnectUnity Core v3.0.6
     - WalletConnectUnity Modal v1.0.6
     - WalletConnectUnity UI v1.0.6

**Is this happening in Unity Editor or on a player like Android, iOS, Windows?**
Unity Editor

**Wallets used for testing**
Metamask Android

**Additional context**
I think it's noteworthy that the "WalletConnectUnity already initialized" error gets logged on every play mode entrance. If I let the WalletConnect.InitializeAsync function return when IsInitialized is true, then the SignClient.Session.Keys is always null. If I comment out the return within the IsInitialized check, then I can proceed to my "store" scene and retrieve signatures from my wallet. If I subsequently return to the main menu and the Modal tries to re-initialize, then I go back to the store scene, then the SignClient.Session.Keys is then null. 

Also, possibly noteworthy, is that even if I delete the folder `/home/[user]/.config/unity3d/[company name]/[game name]/WalletConnect` and/or the storage.json file within, the WalletConnect modal still resumes my sessions on Awake sometimes. Maybe this is leaving my connection in a broken state which is why my Keys would be null?

**Comment by skibitsky** [comment]
Thanks for opening the issue, @underclockeddev.

Have you been able to reproduce it in a player (desktop/mobile)?

**Comment by justynspooner** [comment]
We're having a similar issue in 3.0.6. keep getting exceptions when calling `RequestAsync`

```
var result = await WalletConnect.Instance.RequestAsync<PersonalSign, SignMessageResult>(data);
```

Exception is thrown `WalletConnectSharp.Common.Model.Errors.WalletConnectException`
Message `Something went wrong`

Seems to be a regression in this build?

We're using Desktop Unity 2021.3.15f1

**Comment by skibitsky** [comment]
Hey @justynspooner, your issue might be different. Could you please open a new issue and include the namespaces you’re using?

**Comment by underclockeddev** [comment]
@skibitsky I was able to confirm the issue is the same on an Android player, via console logs with android logcat

**Comment by skibitsky** [comment]
@underclockeddev does this issue occur only with `eth_signTypedData_v4`? If `personal_sign` works correctly, the problem might be in your data structure. Could you please share the definition of your` Permit` class?

**Comment by underclockeddev** [comment]
```
[Struct("Permit")]
protected class Permit
{
    [Parameter("uint256", "deadline", 5)]
    public BigInteger Deadline { get; set; }

    [Parameter("uint256", "nonce", 4)]
    public BigInteger Nonce { get; set; }

    [Parameter("address", "owner", 1)]
    public string Owner { get; set; }

    [Parameter("address", "spender", 2)]
    public string Spender { get; set; }

    [Parameter("uint256", "value", 3)]
    public string Value { get; set; }
}
```

**Comment by underclockeddev** [comment]
I can actually go all the way through a signTypedData call on my first try per app run, so I don't think it's a structuring issue. In fact, the error occurs during `NethereumWalletConnectService.GetWalletConnectConnectedSession()` which is called in all eth & personal function overrides - but only after I revisit the main menu of my game. It seems like possibly I might be trying to reload/reinstantiate in some way the WC framework and it's ruining my configuration?

**Comment by underclockeddev** [comment]
I made some tiny changes to the official WC Unity demo in order to demonstrate my problem, and it totally was reproducible! 

1. Clone my fork https://github.com/underclockeddev/WalletConnectUnity/tree/project/modal-sample
2. Start up the Unity project
3. Enter play mode
4. Connect your wallet
5. Click personal_sign to verify that it works
6. Click the Change Scene button
7. Click the Back to First Scene button
8. Click personal_sign again, and it does not work

**Comment by skibitsky** [comment]
Thanks, @underclockeddev, fantastic work!

I managed to resolve the issue in your project by converting the `instance` field of `WalletConnectContainer` to static.

The problem was that upon returning to the first scene, a non-static instance field would always be empty leading to the creation of second Modal instance.

I'll look into ways to better handle such cases, but for now, `WalletConnectContainer` with static `instance` field should be a good solution.

**Comment by underclockeddev** [comment]
My hero! Thank you so much @skibitsky. Just tested this out and it solved the issue in my actual project too. 

Just to give a little more direction on the issue - I only added the WalletConnectContainer class in order to try and persist the Modal. If you have the Modal prefab itself on the scene, and then switch scenes and switch back, that's problematic as well. Not sure if that's the exact same issue or if I created a new issue by making the container - but that's what I was intending to work around by making the container. 

Thanks again! I'm unblocked :smile:
